connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala

```diff
@@ -113,6 +118,56 @@ class SessionCache(val cacheSize: Int) extends Logging {
           s"Failed to create RiakClient(java) for $endpointsStr", e)
     }
   }
+
+  private case class ReleaseTask(session: CachedSession) extends Runnable {
+    override def run(): Unit = {
+      releaseImmediately(session)
+    }
+  }
+
+  private def releaseDeferred(session: CachedSession, releaseDelayMillis: Long): Unit = {
+      val newTime = System.currentTimeMillis + releaseDelayMillis
+      val releaseTime = deferredReleases.remove(session) match {
+        case Some(oldTime) => math.max(oldTime, newTime)
+        case None          => newTime
+      }
+      deferredReleases.put(session, releaseTime)
+  }
+
+  private def releaseImmediately(session: CachedSession): Unit = {
+    if (session.getRefCount == 1)
+      evict(session.conf)
+  }
+
+  def release(session: CachedSession, timeout: Long = 0): Unit = {
+    if (timeout == 0 || scheduledExecutorService.isShutdown)
+      releaseImmediately(session)
+    else
+      releaseDeferred(session, timeout)
+  }
+
+  private def checkForExpiredSessions(): Unit = {
+    val now = System.currentTimeMillis
+    for ((session, time) <- deferredReleases)
+      if (time <= now && deferredReleases.remove(session).isDefined)
+        ReleaseTask(session).run
+  }
+
+  private val checkForExpiredSessionsTask = new Runnable() {
+    override def run(): Unit = {
+      checkForExpiredSessions
+    }
+  }
+
+  private val scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
```

Comment:
Sorry but from my PoV it isn't.. You are using it for running 1 recurring task in the one thread..

Therefore for me personally, in this particular ca, having the overall processing cycle in the one place is more clearly:
```
() -> {
// Setup thread name
// Do something 
sleep();
}
```